lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -753,24 +804,41 @@ class MongoClientInterface {
         });
     }
 
+    /**
+     * puts object metadata in bucket
+     * @param {String} bucketName bucket name
+     * @param {String} objName object name
+     * @param {Object} objVal object metadata
+     * @param {object} params params
+     * @param {Object} log logger
+     * @param {Function} cb callback
+     * @return {undefined}
+     */
     putObject(bucketName, objName, objVal, params, log, cb) {
         MongoUtils.serialize(objVal);
         const c = this.getCollection(bucketName);
-        if (params && params.versioning && !params.versionId) {
-            return this.putObjectVerCase1(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId === '') {
-            return this.putObjectVerCase2(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId && !params.repairMaster) {
-            return this.putObjectVerCase3(c, bucketName, objName, objVal,
-                params, log, cb);
-        } else if (params && params.versionId && params.repairMaster) {
-            return this.putObjectVerCase4(c, bucketName, objName, objVal,
-                params, log, cb);
-        }
-        return this.putObjectNoVer(c, bucketName, objName, objVal,
-            params, log, cb);
+        const _params = Object.assign({}, params);
+        return this.getBucketVFormat(bucketName, log, (err, vFormat) => {
+            if (err) {
+                return cb(err);
+            }
+            _params.vFormat = vFormat;
+            if (_params && _params.versioning && !_params.versionId) {
```

Comment:
Interestingly, all the following function calls use the same arguments. You could split this into one function that returns the `putObjectStrategy`, in which you could early-return in a lot of cases and improve readability.

```js
function getPutObjectStrategy(params) {
    if (!_params || _params.versionId === undefined) {
        return this.putObjectNoVer;
    }
    if (_params.versionId === '') {
        return this.putObjectVerCase2;
    }
    if (!_params.repairMaster) 
[...]
```